### ISSUE TITLE
Dev: precommit/push git hooks

### ISSUE THREAD

**Issue by riotrah** [BetterThanTomorrow/calva#1653]
I couldn't tell if there were already any (it doesn't appear so).

It would be good prob to introduce (to start with)

- eslint check
- prettier format (`--write`)
- at least mocha unit tests passing

With popular npm package [`husky`](https://www.npmjs.com/package/husky) this typically ends up being _really_ easy.

**Comment by corasaurus-hex** [comment]
There aren't any git hooks currently. I wouldn't be opposed to someone setting up an opt-in git hook setup for themselves but I know that both @PEZ and I are not fans of using these for checks/tests. For me, at least, it feels like a waste to have devs spend time waiting for these when we have CI to make certain these things are done. For some people it's worth it and I totally understand how that goes and so I'd be ok with something we can opt in to if we want but is off by default.

Thoughts? Maybe have an env var you can enable? Or a file you can touch in your local .git directory that a hook checks for the existence of before running checks?

**Comment by corasaurus-hex** [comment]
here's some prior art on using the .git directory to enable certain features local to just your own checkout https://thoughtbot.com/blog/git-safe

**Comment by PEZ** [comment]
TL;DR: I think we can skip this.

My uses cases for this are covered by the new build task. My use cases are:

1. I won't discover later that formatting was off
    * Covered by the watcher that formats code all the time 
2. I won't discover later that linting was off
   * Covered by the watcher that shows me if something didn't please the linter
3. I won't discover later that some unit test doesn't pass
   * Covered by ty the CLJS and TS(*) test watchers

As for **3** I actually sometimes want to commit failing unit tests.

(*) The TS test watcher is broken currently. It doesn't seem to pick up new code.

**Comment by riotrah** [comment]
Interesting about the git safe stuff!

@PEZ these are good points.

Perhaps just adding a lint check or write step to `compile` or `precompile` would suffice. I seem to tend to change files before I remember to run the watchers, so format failures only get caught in cdci alas

**Comment by PEZ** [comment]
I've added a pre-step to the formatting watcher now that first formats everything.
